### PR TITLE
Single ac binary

### LINKED ISSUE #82: One binary to rule them all
We may lower the size of AppController docker image and decrease image build times by mergin all our commands into single binary.

We would need a top-level `ac` binary, which would have subcommands

- `ac wrap`
- `ac deploy`
- `ac bootstrap`

### PR DESCRIPTION
**Pull request by nebril**
Cobra library is used to create one binary with subcommands.
Updated Dockerfile to build just one binary

Updated examples to use wrap command from single binary

fixes #82

---
This change is [Reviewable](https://reviewable.io/reviews/mirantis/k8s-appcontroller/83)

### CODE REVIEW

File: manifests/appcontroller.yaml [hunk 1]
```diff
@@ -3,10 +3,10 @@ kind: Pod
 metadata:
   name: k8s-appcontroller
   annotations:
-    pod.alpha.kubernetes.io/init-containers: '[{"name": "kubeac-bootstrap", "image": "mirantis/k8s-appcontroller", "command": ["bootstrap", "/opt/kubeac/manifests"]}]'
+    pod.alpha.kubernetes.io/init-containers: '[{"name": "kubeac-bootstrap", "image": "mirantis/k8s-appcontroller:single-binary", "command": ["kubeac", "bootstrap", "/opt/kubeac/manifests"]}]'
```
Comment (gluke77) [inline manifests/appcontroller.yaml]: If we are to merge it to master, we should not point to tagged docker image but use trunk

File: manifests/appcontroller.yaml [hunk 2]
```diff
@@ -3,10 +3,10 @@ kind: Pod
...
 spec:
   restartPolicy: Always
   containers:
   - name: kubeac
-    image: mirantis/k8s-appcontroller
+    image: mirantis/k8s-appcontroller:single-binary
```
Comment (gluke77) [inline manifests/appcontroller.yaml]: Ditto

File: manifests/appcontroller.yaml (hunk 1 above)
Comment (jellonek) [inline manifests/appcontroller.yaml]: Seems like remains from development, am I correct?

File: main.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+	"github.com/Mirantis/k8s-AppController/cmd"
+)
+
+func main() {
+	cmd.Init()
+
+	if err := cmd.RootCmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(-1)
```
Comment (gluke77) [inline main.go]: That looks too hard. Why don't just log.Fatal ?

**Comment by pigmej** [comment]
Review status: 0 of 17 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[ac_service.sh, line 7 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83#-KSG9ACM5qGmonMWTOSH:-KSG9ACM5qGmonMWTOSI:bsfbuug) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/f479b19d1899bb185dfb6021ff117ee85f0d9b95/ac_service.sh#L7)):*
> deploy

I dislike that name, why deploy?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83)*

File: Dockerfile [hunk 4]
```diff
@@ -6,7 +6,7 @@ COPY . /go/src/github.com/Mirantis/k8s-AppController
 WORKDIR /go/src/github.com/Mirantis/k8s-AppController
 
 RUN apk --no-cache add git
-RUN go get ./... && go build -o kubeac cmd/kubeac/main.go && go build -o wrap cmd/wrap/main.go && go build -o bootstrap cmd/bootstrap/main.go && mv kubeac /usr/bin/kubeac && mv wrap /usr/bin/wrap && mv bootstrap /usr/bin/bootstrap && mkdir -p /opt/kubeac && mv /go/src/github.com/Mirantis/k8s-AppController/manifests /opt/kubeac/manifests && rm -fr /go
+RUN go get ./... && go build -o kubeac && mv kubeac /usr/bin/kubeac && mkdir -p /opt/kubeac && mv /go/src/github.com/Mirantis/k8s-AppController/manifests /opt/kubeac/manifests && rm -fr /go
```
Comment (jellonek) [inline Dockerfile]: :+1:

File: manifests/appcontroller.yaml (hunk 1 above)
Comment (nebril) [inline manifests/appcontroller.yaml]: indeed, sorry about that, thanks for catching it.

File: main.go (hunk 3 above)
Comment (nebril) [inline main.go]: sure, makes sense

**Comment by nebril** [comment]
Review status: 0 of 17 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[ac_service.sh, line 7 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83#-KSG9ACM5qGmonMWTOSH:-KSGQvfEPRH1avBRhfA2:b-3792g7) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/f479b19d1899bb185dfb6021ff117ee85f0d9b95/ac_service.sh#L7)):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> > deploy
> 
> I dislike that name, why deploy?
</details>
because it deploys the graph. What else would you suggest?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83)*

**Comment by pigmej** [comment]
Review status: 0 of 17 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[ac_service.sh, line 7 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83#-KSG9ACM5qGmonMWTOSH:-KSGUgQYuja6YLXohgE3:b-ip7ra6) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/f479b19d1899bb185dfb6021ff117ee85f0d9b95/ac_service.sh#L7)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> because it deploys the graph. What else would you suggest?
</details>
You need to think in advance, what when we will make AC "running forever" ? What if we will support that "live" case? It should be `start-ac` or something like that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83)*

**Comment by nebril** [comment]
Review status: 0 of 17 files reviewed at latest revision, 5 unresolved discussions.

---

*[ac_service.sh, line 7 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83#-KSG9ACM5qGmonMWTOSH:-KSGX_9M3xHPzZhNb4AA:b-o34g3z) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/f479b19d1899bb185dfb6021ff117ee85f0d9b95/ac_service.sh#L7)):*
<details><summary><i>Previously, pigmej (Jędrzej Nowak) wrote…</i></summary>
> You need to think in advance, what when we will make AC "running forever" ? What if we will support that "live" case? It should be `start-ac` or something like that.
</details>
We should change the name of the command when this happens. Right now this command is deploying the graph, and it's exactly what the name states it does.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83)*

**Comment by nebril** [comment]
Review status: 0 of 17 files reviewed at latest revision, 5 unresolved discussions.

---

*[Dockerfile, line 9 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83#-KSG75I0vym95ExUyvj8-r1-9:-KSGY2aN_Fh7D-GgUyP-:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/f479b19d1899bb185dfb6021ff117ee85f0d9b95/Dockerfile#L9)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary>
> :+1:
</details>
Done.

---

*[main.go, line 29 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83#-KSG75IIZNj6gskcvx17-r1-29:-KSGY3PVcg9R1OFl77e8:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/f479b19d1899bb185dfb6021ff117ee85f0d9b95/main.go#L29)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> sure, makes sense
</details>
Done.

---

*[manifests/appcontroller.yaml, line 6 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83#-KSG75IIZNj6gskcvx0x-r1-6:-KSGY48ARXvqdvbRn8Ho:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/f479b19d1899bb185dfb6021ff117ee85f0d9b95/manifests/appcontroller.yaml#L6)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> +1, forgot about it, thanks for catching
</details>
Done.

---

*[manifests/appcontroller.yaml, line 11 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83#-KSG75IIZNj6gskcvx0x-r1-11:-KSGY4Z5taPjTWpCQv_3:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/f479b19d1899bb185dfb6021ff117ee85f0d9b95/manifests/appcontroller.yaml#L11)):*
<details><summary><i>Previously, gluke77 (Illia Khudoshyn) wrote…</i></summary>
> Ditto
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83)*

**Comment by nebril** [comment]
Review status: 0 of 17 files reviewed at latest revision, 5 unresolved discussions.

---

*[ac_service.sh, line 7 at r1](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83#-KSG9ACM5qGmonMWTOSH:-KSGnMjALFgioJo3OBJm:b-896fix) ([raw file](https://github.com/mirantis/k8s-appcontroller/blob/f479b19d1899bb185dfb6021ff117ee85f0d9b95/ac_service.sh#L7)):*
<details><summary><i>Previously, nebril (Maciej Kwiek) wrote…</i></summary>
> We should change the name of the command when this happens. Right now this command is deploying the graph, and it's exactly what the name states it does.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83)*

**Comment by pigmej** [comment]
:lgtm:

---

Reviewed 13 of 26 files at r1, 4 of 5 files at r2, 9 of 9 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/k8s-appcontroller/83#-:-KSGoQtBcHvnULnpBwhm:bnfp4nl)*

**Pull request merged**